lib/planner/search-algorithms/action-tree.js

```diff
@@ -0,0 +1,954 @@
+        let valueIndex = -1;
+        if (actionTree._unmetPreconditions.hasOwnProperty(precon.key)) {
+          let unmetPrecon = actionTree._unmetPreconditions[precon.key];
+          for (let j = 0; j < unmetPrecon.values.length; j++) {
+            let matchingPrecon = true;
+            try {
+              assert.deepEqual(precon.value, unmetPrecon.values[j], true);
+            } catch (err) {
+              matchingPrecon = false;
+            }
+            if (matchingPrecon) {
+              valueIndex = j;
+              break;
+            }
+          }
+        } else {
+          possible = false;
+          break;
+        }
+
+        // Check to see if the there is a setter for a valid valueIndex
+        if (valueIndex > -1 && actionTree._unmetPreconditions[precon.key].setters[valueIndex].length > 0) {
+          nodesToHit.push(actionTree._unmetPreconditions[precon.key].setters[valueIndex]);
+          valueIndexs.push(valueIndex);
+        } else {
+          possible = false;
+          break;
+        }
+      }
+
+      // If all preconditions are met for this particlar node, create the path
+      if (possible) {
+        let nodePath = yield actionTree.emitAsync('actionTree.constructPath', node, nodesToHit, next);
+        if (nodePath.length > 0) {
+          // The node we found a path to can be removed from preconmap
+          actionTree._removeFromPreconMap(node, allPrecons, valueIndexs);
+          // Find its correct new home in the tree and add to actionStack
+          actionTree._processNewPath(nodePath);
+          // Add the node to the action stack with its new path
+          actionTree._actionStack.push(nodePath[nodePath.length - 1].id);
+        }
+      }
+    }
+    return callback(null);
+  },
+  _processNewPath(nodePath) {
+    let i = 0;
+    for (i; i < nodePath.length; i++) {
+      if (nodePath[i].parent) {
+        let flag = true;
+        for (let child of nodePath[i - 1].children) {
+          if (child.id === nodePath[i].id) {
+            flag = false;
+            break;
+          }
+        }
+        if (flag) {
+          break;
+        }
+      }
+    }
+    /*
+      TODO: Once nodes start getting cloned below,
+      they are making unique actions in a unique state that they were not previously doing
+      Do we want to take account for this,? check if it effects the state in a new way?
+      Adding to metaES and unmetPrecons
+      Not sure if its wanted, as this may lead to false positives that mutiple actions can solve one precondition.
+    */
+
+    for (i; i < nodePath.length - 1; i++) {
+      let node = actionTree._createNode(nodePath[i], nodePath[i].action, nodePath[i - 1], true);
+      nodePath[i] = node;
+    }
+
+    nodePath[nodePath.length - 1].parent = nodePath[nodePath.length - 2];
+    nodePath[nodePath.length - 2].children.push(nodePath[nodePath.length - 1]);
+
+    // // Travel down the path, until we find a node out of place
```

Comment:
Delete commented code. It is in source control we'll find it if we need it again